synapse/handlers/_base.py

```diff
@@ -56,41 +85,39 @@ def _filter_events_for_client(self, user_id, events):
             )
         )
 
-        def allowed(event, state):
-            if event.type == EventTypes.RoomHistoryVisibility:
-                return True
+        events_to_return = []
+        events_to_return_if_later_joined = []
+        for event in events:
+            state = event_id_to_state[event.event_id]
 
-            membership_ev = state.get((EventTypes.Member, user_id), None)
-            if membership_ev:
-                membership = membership_ev.membership
+            membership_event = state.get((EventTypes.Member, user_id), None)
+            if membership_event:
+                membership = membership_event.membership
             else:
-                membership = Membership.LEAVE
+                membership = None
 
             if membership == Membership.JOIN:
-                return True
+                events_to_return.extend(events_to_return_if_later_joined)
+                events_to_return_if_later_joined = []
 
-            history = state.get((EventTypes.RoomHistoryVisibility, ''), None)
-            if history:
-                visibility = history.content.get("history_visibility", "shared")
+            visibility_event = state.get((EventTypes.RoomHistoryVisibility, ""), None)
+            if visibility_event:
+                visibility = visibility_event.content.get("history_visibility", "shared")
             else:
                 visibility = "shared"
 
-            if visibility == "public":
-                return True
-            elif visibility == "shared":
-                return True
-            elif visibility == "joined":
-                return membership == Membership.JOIN
-            elif visibility == "invited":
-                return membership == Membership.INVITE
+            should_include = allowed(event, membership, visibility)
+            if should_include == include:
+                events_to_return.append(event)
+            elif should_include == include_if_later_joined:
+                events_to_return_if_later_joined.append(event)
 
-            return True
+        if is_guest and len(events_to_return) < len(events):
```

Comment:
This might need a comment to explain why ``len(events_to_return) < len(events)`` means that a guest is forbidden.